compiler/passes/src/type_checking/check_statements.rs

```diff
@@ -286,10 +288,32 @@ impl<'a> StatementVisitor<'a> for TypeChecker<'a> {
             // Note that this check is needed because the pass attempts to make progress, even though the literal may be invalid.
             if let Ok(value) = Value::try_from(literal) {
                 input.stop_value.replace(Some(value));
+            } else {
+                self.emit_err(TypeCheckerError::loop_bound_must_be_a_literal(input.stop.span()));
             }
         } else {
             self.emit_err(TypeCheckerError::loop_bound_must_be_a_literal(input.stop.span()));
         }
+
+        // Ensure loop bounds are not decreasing.
+        if match (input.start_value.borrow().as_ref(), input.stop_value.borrow().as_ref()) {
+            (Some(Value::I8(lower_bound, _)), Some(Value::I8(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::I16(lower_bound, _)), Some(Value::I16(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::I32(lower_bound, _)), Some(Value::I32(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::I64(lower_bound, _)), Some(Value::I64(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::I128(lower_bound, _)), Some(Value::I128(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::U8(lower_bound, _)), Some(Value::U8(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::U16(lower_bound, _)), Some(Value::U16(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::U32(lower_bound, _)), Some(Value::U32(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::U64(lower_bound, _)), Some(Value::U64(upper_bound, _))) => lower_bound >= upper_bound,
+            (Some(Value::U128(lower_bound, _)), Some(Value::U128(upper_bound, _))) => lower_bound >= upper_bound,
+            _ => {
+                self.emit_err(TypeCheckerError::loop_bound_type_mismatch(input.stop.span()));
```

Comment:
Creates spurious type errors.
Check `loop_non_literal_bound_fail.leo` and `loop_non_literal_bound_fail.leo.out`.